RCode/NEFI/partition_uncertainty_example-RandWalk-Year.R

```diff
@@ -0,0 +1,247 @@
+  ecoforecastR::ciEnvelope(cal_time, ci[1,], ci[3,], col="lightBlue")
+  lines(cal_time, ci[2,], col="blue")
+  points(cal_time, ci[2,])
+}
+
+ci <- apply(exp(out[ ,mus]), 2, quantile, c(0.025,0.5,0.975))
+plot.run()
+
+
+
+#setting up deterministic forecast function for random walk with random year
+# IC = initial conditions from end of calibration period 
+# N_out = number of time steps to forecast 
+# tau_add = process error (defaults to zero) 
+# tau_yr = random effect error? (also defaults to zero?)
+# n = number of monte carlo sims 
+forecast_gloeo <- function(IC, N_out = N_out, tau_add = 0, tau_yr = 0, n = Nmc){
+  out <- matrix(NA, n, N_out) # setting up output 
+  gloeo_prev <- IC
+  for(i in 1:N_out){
+    out[,i] <- rnorm(n, gloeo_prev, tau_add) # predict next step JB getting error here when include tau_yr - unsure why?
+    gloeo_prev <- out[,i] # update IC 
+  }
+  return(out)
+}
+
+## deterministic predictions 
+N_out = nrow(dat$forecast) # length of forecast time points 
+IC = mu
+
+gloeo.det <- forecast_gloeo(IC = mean(IC[,N]), # last column of cal is the initial condition for forecast 
+                            N_out = N_out,
+                            tau_add = 0,
+                            tau_yr = 0,
+                            n = 1)
+
+## Plot run
+plot.run()
+lines(forecast_time, gloeo.det, col="purple", lwd=3)
+
+
+######## initial condition uncertainty #######
+# sample rows from previous analysis
+prow = sample.int(nrow(out),Nmc,replace=TRUE)
+
+gloeo.I <- forecast_gloeo(IC = IC[prow, N],
+                      N_out = N_out,
+                      tau_add = 0,
+                      tau_yr = 0,
+                      n = Nmc)
+
+## Plot run
+plot.run()
+gloeo.I.ci = apply(gloeo.I, 2, quantile, c(0.025,0.5,0.975))
+ecoforecastR::ciEnvelope(forecast_time, gloeo.I.ci[1,], gloeo.I.ci[3,], col = col.alpha(N.cols[1], trans))
+lines(forecast_time, gloeo.I.ci[2,], lwd=0.5)
+
+###### parameter uncertainty #######
+# we don't have this for random walk 
+
+###### driver uncertainty ########## 
+# we don't have this for random walk 
+
+###### process uncertainty ######### 
+tau_add_mc <- 1/sqrt(out[prow,"tau_add"])  ## converts from precision to standard deviation
+
+gloeo.IP <- forecast_gloeo(IC = IC[prow, N],
+                          N_out = N_out,
+                          tau_add = tau_add_mc,
+                          tau_yr = 0,
+                          n = Nmc)
+
+## Plot run
+plot.run()
+gloeo.IP.ci = apply(gloeo.IP, 2, quantile, c(0.025,0.5,0.975))
+ecoforecastR::ciEnvelope(forecast_time, gloeo.IP.ci[1,], gloeo.IP.ci[3,], col = col.alpha(N.cols[2], trans))
+ecoforecastR::ciEnvelope(forecast_time, gloeo.I.ci[1,], gloeo.I.ci[3,], col = col.alpha(N.cols[1], trans))
+lines(forecast_time, gloeo.I.ci[2,], lwd=0.5)
+
+###### randome effect uncertainty #######
+
+## Random effect samples
+tau_yr_mc <- 1/sqrt(out[prow,"tau_yr"]) ## converts from precision to std deviation
+aNew.mc <- rnorm(Nmc,0,tau_yr_mc)
```

Comment:
we want to preserve covariance among initial conditions, states, and parameters so we should sample from the posterior of our fitted model. See suggestion above for sampling from the last year's fitted random effect